src/projects/lja/multi_graph.hpp

```diff
@@ -293,6 +377,106 @@ namespace multigraph {
+                    return result_map;
+                int rcid = vertices[vid]->rc->id;
+                std::set<int> edgeids_to_remove;
+                Edge* e_out =  vertices[vid]->outgoing[0];
+                Vertex* end_v = e_out->end;
+                edgeids_to_remove.insert(e_out->getId());
+//                edgeids_to_remove.insert(e_out->rc->id)
+                Edge* e_in = vertices[vid]->rc->outgoing[0]->rc;
+                Vertex* start_v = e_in->start;
+                edgeids_to_remove.insert(e_in->getId());
+//                edgeids_to_remove.insert(e_out->rc->id);
+                size_t overlap = vertices[vid]->k();
+                size_t pref = e_in->getSeq().size();
+                if (pref >= overlap )
+                        pref -= overlap;
+                else {
+                    pref = 0;
+                    std::cerr << " wrong overlap " << overlap << " " << e_in->getSeq().size() << endl;
+                }
+                Sequence new_seq = e_in->getSeq().Prefix(pref) + e_out->getSeq();
+                string new_label = e_in->getLabel()+ "_"+ e_out->getLabel();
+/*                if (haplo_map_ != nullptr) {
+                    HaplotypeStats new_haplo(new_label, (haplo_map_->find(e_in->getLabel())->second),
+                                             (haplo_map_->find(e_out->getLabel()))->second);
+                    haplo_map_->insert(make_pair(new_label, new_haplo));
+//                    logger.info() << new_label << endl;
+                } */
+
+                result_map[new_label] = {e_in->getLabel(), e_out->getLabel()};
+                addEdge(*start_v, *end_v, new_seq, 0, new_label);
+                for (auto eid: edgeids_to_remove){
+                    internalRemoveEdge(eid);
+                }
+                
+                delete vertices[vid];
+                delete vertices[rcid];
+                vertices.erase(vid);
+                vertices.erase(rcid);
+            }
+            return result_map;
+        }
+
+        deleted_edges_map deleteEdgeById(int eid){
+            Vertex* start_v = edges[eid]->start;
+            Vertex* end_v = edges[eid]->end;
+            internalRemoveEdge(eid);
+
+            auto res = compressVertex(start_v->id);
+
+//this compression may contain edges that result from first one, so additional ugly processing required.
+            auto comp_res = compressVertex(end_v->id);
+            for (auto p: comp_res) {
+                std::vector<std::string> patched_old;
+                for (auto comp_edge: p.second) {
+                    if (res.find(comp_edge) != res.end()) {
+                        patched_old.insert(patched_old.end(), res[comp_edge].begin(), res[comp_edge].end());
+                    } else
+                        patched_old.push_back(comp_edge);
+                }
+                res[p.first] = patched_old;
+            }
+            return res;
+        }
+
```

Comment:
Agree, this can be optimized. However, on HG002 it is really fast (5 minutes).
Seems that in real case with random collapsing order (i.e. because of the unordered_map) this is more like  O(genome length) * O(log(edge number))

Reply:
Think a bit more, see Merge in multigraph